src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -400,82 +523,171 @@ impl<O: ForestObligation> ObligationForest<O> {
-            // happen, and code in `compress` has to allow for it.
-            if node.state.get() != NodeState::Pending {
-                index += 1;
-                continue;
-            }
+        self.unblock_nodes(processor);
+
+        let mut run_again = true;
+        while !self.unblocked.is_empty() || (!self.check_next.is_empty() && run_again) {
+            run_again = false;
+            let nodes = &self.nodes;
+            self.unblocked.extend(
+                self.check_next
+                    .drain(..)
+                    .map(|index| Unblocked { index, order: nodes[index].node_number }),
+            );
+            while let Some(Unblocked { index, .. }) = self.unblocked.pop() {
+                if self.unblocked.peek().map(|u| u.index) == Some(index) {
+                    continue;
+                }
+                let node = &mut self.nodes[index];
 
-            match processor.process_obligation(&mut node.obligation) {
-                ProcessResult::Unchanged => {
-                    // No change in state.
+                if node.state.get() != NodeState::Pending {
+                    continue;
                 }
-                ProcessResult::Changed(children) => {
-                    // We are not (yet) stalled.
-                    stalled = false;
-                    node.state.set(NodeState::Success);
-
-                    for child in children {
-                        let st = self.register_obligation_at(child, Some(index));
-                        if let Err(()) = st {
-                            // Error already reported - propagate it
-                            // to our node.
-                            self.error_at(index);
+                for var in node.obligation.stalled_on() {
+                    match self.stalled_on.entry(var.clone()) {
+                        Entry::Vacant(_) => (),
+                        Entry::Occupied(mut entry) => {
+                            let nodes = entry.get_mut();
+                            nodes.retain(|node_index| *node_index != index);
+                            if nodes.is_empty() {
+                                processor.unwatch_variable(var.clone());
+                                entry.remove();
+                            }
                         }
                     }
                 }
-                ProcessResult::Error(err) => {
-                    stalled = false;
-                    errors.push(Error { error: err, backtrace: self.error_at(index) });
+
+                // `processor.process_obligation` can modify the predicate within
+                // `node.obligation`, and that predicate is the key used for
+                // `self.active_cache`. This means that `self.active_cache` can get
+                // out of sync with `nodes`. It's not very common, but it does
+                // happen, and code in `compress` has to allow for it.
```

Comment:
Unfortunately not, this is entirely out of control of `ObligationForest` since it is the implementation of `process_obligation` which does the mutation. (I had to compensate for this with the `alternative_predicates` addition.